feat(sources): ds-503 add hosts field validation

mirekdlugosz opened the pull request:
This was originally part of #475 .

Add function that validates the content of hosts field. It is generally meant to reject empty lists and lists that have too many items (everything but network works against single host), but it also looks into individual hosts and rejects ones that won't be accepted by backend anyway.

Regular expressions used in `hostValid` are [copied from backend](https://github.com/quipucords/quipucords/blob/main/quipucords/api/source/serializer.py#L329).

One thing I don't like in current implementation is that field is marked as invalid right as you open the modal. It should be marked as such only if user touched a field / tries to submit form without touching a field. Hints how to solve that are most appreciated!

cdcabrera commented on src/helpers/helpers.ts:
```diff
@@ -111,6 +112,50 @@ const normalizeHosts = (data?: string) =>
+ * @param {number} maxHosts - maximum number of hosts that the field should allow.
+ * @returns ValidatedOptions
+ */
+const validateHosts = (data: string | undefined, maxHosts: number) => {
```
required: lets use a default for `maxHosts`. reason... old skool js, secondary params are associated with support settings and we can provide a tiny bit of developer "ease of use"

up to you whether you want to use `maxHosts: number = Infinity` or `maxHosts: number = 1` (i partially assume it would default to Infinity)

once you do this your typescript can also be updated towards (and of course the call where you declare `Infinity` in `addSourceModal` can be updated)

```
const validateHosts = (data?: string, maxHosts: number = Infinity) => {
```

mirekdlugosz replied:
Done c8a96d131f94ebd901a405ef37d2ca19ebca73a5 .

I still supply both arguments on caller side, to make it more explicit that there are different rules based on source type.

cdcabrera commented on src/helpers/helpers.ts:
```diff
@@ -111,6 +112,50 @@ const normalizeHosts = (data?: string) =>
+
+  const normalizedHosts = normalizeHosts(data);
+
+  if (!normalizedHosts!.length) {
```
looks like we forgot the eslint rule `'@typescript-eslint/no-non-null-assertion'` that's on me... apologies. it's difficult to read and there's the baked in JS optional chaining

**required:** all instances of this "non-null-assertion", `!.` should be replaced... prefer optional chaining, `?.`, instead. that would make this line

```
if (!normalizedHosts?.length)
```
and of course update the remaining instances of this

**optional**
add this rule to the eslint config to avoid this scenario again
```
'@typescript-eslint/no-non-null-assertion': 2,
```

mirekdlugosz replied:
Done in c8a96d131f94ebd901a405ef37d2ca19ebca73a5 , along with optional.

I didn't like this `variable!`, and felt it is not necessary. In cbfbcd6faa3b010f2204e122283e436d06a328ee I took another look and modified `normalizeHosts` in a way that makes both `?.` and `!.` not necessary.

This is another commit so we can easily drop it if there are any problems with that approach.

cdcabrera commented on src/helpers/__tests__/helpers.test.ts:
```diff
@@ -155,6 +156,46 @@ describe('normalizeHosts', () => {
+    ['my-rhel[a:d].company.com'],
+    ['my-rhel[120:400].company.com'],
+    ['my-rh_el[120:400].comp_a-ny.com']
+  ])('should accept valid host [%s]', host => {
```
sweet use! good call!

mirekdlugosz replied:
Thanks :smiley:

cdcabrera reviewed (changes requested):
Almost there! Some old Typescript we should squash

mirekdlugosz commented:
On top of fixes for things pointed out in review, I introduced a bit larger change in cbfbcd6faa3b010f2204e122283e436d06a328ee .

It ensures that `normalizeHosts` returns array, not array or undefined. That makes things a bit more readable in `validateHosts`, but the cost is that it is no longer pure arrow function. Not sure if that matters.

I also changed default SourceFormType state to have `hosts` as `undefined` (instead of empty string). This allows us to differentiate between "field is empty because modal has been just opened" (field not marked as invalid) and "field is empty because user typed something and removed it" (field marked as invalid). `normalizeHosts` still changes that to empty list, and we could prevent form submission once we have that code.

However, I am not sure if that `undefined` will not cause any issues elsewhere, so I am open to dropping this commit.

cdcabrera reviewed (approved):
tested using similar regular expressions used in unit tests. appears to function as intended

>there was some history around the specific old limits setup for subnets... unclear if that's changed or because of the openness of the new checks/currently missing validation, if that's still a concern... regardless it can be iterated on
>
> - https://github.com/quipucords/quipucords-ui/blob/1.8.0/src/components/addSourceWizard/__tests__/addSourceWizardStepTwo.test.js#L98-L116

mirekdlugosz commented:
In c8baf9da1c682f660d4831cd6aaabeca837e9e02 I added some strings from old unit tests. I skipped following:

* `0.0.0.0:22` this is indeed invalid, but gets through new validation and I don't think I can catch it without making that code much more complex. Since the long term goal is to perform all validation on server side, I think we can ignore it for now.
* `0.0.0.12/15` this used to be considered invalid, but I have no idea why. This is perfectly valid [CIDR block](https://en.wikipedia.org/wiki/Classless_Inter-Domain_Routing#IPv4_CIDR_blocks). It will result in a scan of over 130 000 hosts, which is likely not what anyone needs, but technically it is correct and I don't think we can know better than the user what they want.
* `0.0.0.12/33` this is indeed invalid, but again - we would have to use more complex code to catch it. For now we allow any two-digit number.
